---
title: "Semen microbiome - Species Diversity Plots (Figure 1)"
output: html_notebook
---

```{r "setup", include=FALSE}
require("knitr")
opts_knit$set(root.dir='./')
```

Library Imports
```{r}
library(ggplot2)
library(readr)
library(ggsignif)
library(scales)
library(ggpubr)
library(rstatix)
library(emmeans)

dir.create('./Analysis Figures/Figure1_subplots/', recursive = TRUE)
dir.create('./Results/Figure1_MultipleComparisons/', recursive = TRUE)
```
Data loading
```{r}
diversityMeasures <- read_csv('../Data/Alpha_Diversity_Measures_ASVs.csv')

cstAssignments <- read_csv('../Code/Results/CST_Analysis/Semen_CST_GeneraAssignments.csv')

diversityMeasures <- merge(diversityMeasures, cstAssignments, by.x='Sample Name', by.y='SampleID')

clinicalData <- read_csv('../Data/ClinicalData.csv')

diversityMeasures <- merge(diversityMeasures, clinicalData, by='Sample Name')
```

# Plots
```{r}
# Color palette from matplotlib/seaborn for reference
# mpl_color_palette <- c('#e41a1c', '#377eb8', '#4daf4a')
lmModel <- lm(shannon_entropy ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

write.csv(contrastRes, './Results/Figure1_MultipleComparisons/Shannon_CST_contrasts.csv')

pwc <- diversityMeasures %>% 
  dunn_test(shannon_entropy ~ CST, p.adjust.method = "holm") 

plotShannon <- ggplot(diversityMeasures, aes(x=as.factor(CST), y=`shannon_entropy`)) + geom_boxplot(aes(col=as.factor(CST)), outlier.shape=NA, alpha=1) + geom_jitter(aes(col=as.factor(CST))) + xlab('Cluster') + theme_classic() +    stat_compare_means(method = "wilcox.test",  comparisons = list(c("1", "2"), c("2", "3"), c("1", "3")), map_signif_level = TRUE, label = "p.signif", p.adjust.methods='holm') +
  scale_color_manual(name='Cluster', values = c("1" = "#e41a1c", "2"="#377eb8", "3"="#4daf4a")) + ylab('Shannon entropy') +
  theme(axis.text.x = element_text(size=16), axis.text.y = element_text(size=16), axis.title.x = element_text(size=18), axis.title.y = element_text(size=18)) 

lmModel <- lm(shannon_entropy ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

lmModel <- lm(observed_features ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

write.csv(contrastRes, './Results/Figure1_MultipleComparisons/Richness_CST_contrasts.csv')

plotRichness <- ggplot(diversityMeasures, aes(x=as.factor(CST), y=`observed_features`)) + geom_boxplot(aes(col=as.factor(CST)), outlier.shape=NA, alpha=1) + geom_jitter(aes(col=as.factor(CST))) + xlab('Cluster') + theme_classic() + 
  geom_pwc(method = "wilcox.test", label = "p.signif", hide.ns = TRUE, p.adjust.method='holm') + 
  #stat_compare_means(method = "wilcox.test",  comparisons = list(c("1", "2"), c("2", "3"), c("1", "3")), map_signif_level = TRUE, label = "p.signif", p.adjust.methods='holm') +
  scale_color_manual(name='Cluster', values = c("1" = "#e41a1c", "2"="#377eb8", "3"="#4daf4a")) + ylab('Number of observed ASVs') +
  theme(axis.text.x = element_text(size=16), axis.text.y = element_text(size=16), axis.title.x = element_text(size=18), axis.title.y = element_text(size=18)) 

lmModel <- lm(faith_pd ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

write.csv(contrastRes, './Results/Figure1_MultipleComparisons/FaithPD_CST_contrasts.csv')

plotFaithPD <- ggplot(diversityMeasures, aes(x=as.factor(CST), y=`faith_pd`)) + geom_boxplot(aes(col=as.factor(CST)), outlier.shape=NA, alpha=1) + geom_jitter(aes(col=as.factor(CST))) + xlab('Cluster') + theme_classic() + 
  geom_pwc(method = "wilcox.test", label = "p.signif", hide.ns = TRUE, p.adjust.method='holm') + 
  # stat_compare_means(method = "wilcox.test",  comparisons = list(c("1", "2"), c("2", "3"), c("1", "3")), map_signif_level = TRUE, label = "p.signif", p.adjust.methods='holm', hide.ns = TRUE)  +
  scale_color_manual(name='Cluster', values = c("1" = "#e41a1c", "2"="#377eb8", "3"="#4daf4a"))    + ylab('Faith\'s phylogenetic diversity') + guides(color='none') +
  theme(axis.text.x = element_text(size=16), axis.text.y = element_text(size=16), axis.title.x = element_text(size=18), axis.title.y = element_text(size=18)) 

lmModel <- lm(simpson ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

write.csv(contrastRes, './Results/Figure1_MultipleComparisons/Simpson_CST_contrasts.csv')

plotSimpson <- ggplot(diversityMeasures, aes(x=as.factor(CST), y=`simpson`)) + geom_boxplot(aes(col=as.factor(CST)), outlier.shape=NA, alpha=1) + geom_jitter(aes(col=as.factor(CST))) + xlab('Cluster') + theme_classic() + 
    geom_pwc(method = "wilcox.test", label = "p.signif", hide.ns = TRUE, p.adjust.method='holm') + 
# stat_compare_means(method = "wilcox.test",  comparisons = list(c("1", "2"), c("2", "3"), c("1", "3")), map_signif_level = TRUE, label = "p.signif", p.adjust.methods='holm') +
  scale_color_manual(name='Cluster', values = c("1" = "#e41a1c", "2"="#377eb8", "3"="#4daf4a")) + ylab('Simpson\'s index') + guides(color='none') +
  theme(axis.text.x = element_text(size=16), axis.text.y = element_text(size=16), axis.title.x = element_text(size=18), axis.title.y = element_text(size=18)) 

lmModel <- lm(simpson_e ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

write.csv(contrastRes, './Results/Figure1_MultipleComparisons/SimpsonEveness_CST_contrasts.csv')

plotSimpson_e <- ggplot(diversityMeasures, aes(x=as.factor(CST), y=`simpson_e`)) + geom_boxplot(aes(col=as.factor(CST)), outlier.shape=NA, alpha=1) + geom_jitter(aes(col=as.factor(CST))) + xlab('Cluster') + theme_classic() + 
   geom_pwc(method = "wilcox.test", label = "p.signif", hide.ns = TRUE, p.adjust.method='holm') + 
  # stat_compare_means(method = "wilcox.test",  comparisons = list(c("1", "2"), c("2", "3"), c("1", "3")), map_signif_level = TRUE, label = "p.signif", p.adjust.methods='holm') +
  scale_color_manual(name='Cluster', values = c("1" = "#e41a1c", "2"="#377eb8", "3"="#4daf4a")) + ylab('Simpson\'s evenness') + guides(color='none') +
  theme(axis.text.x = element_text(size=16), axis.text.y = element_text(size=16), axis.title.x = element_text(size=18), axis.title.y = element_text(size=18)) 


lmModel <- lm(log(`Total 16s rRNA copy per mL`) ~ as.factor(CST), data=diversityMeasures)
eMeansObj <- emmeans(lmModel, ~ as.factor(CST))
contrastRes <- pairs(eMeansObj)

write.csv(contrastRes, './Results/Figure1_MultipleComparisons/16SCopy_CST_contrasts.csv')

plotBacterialAmount <- ggplot(diversityMeasures, aes(x=as.factor(CST), y=`Total 16s rRNA copy per mL`)) + geom_boxplot(aes(col=as.factor(CST)), outlier.shape=NA, alpha=1) + geom_jitter(aes(col=as.factor(CST))) + xlab('Cluster') + theme_classic() +  stat_compare_means(method = "wilcox.test",  comparisons = list(c("1", "2"), c("2", "3"), c("1", "3")), map_signif_level = TRUE, label = "p.signif", p.adjust.methods='holm') + ylab(expression(log[10]~ "(16S rRNA copy number / mL semen)")) + 
 scale_color_manual(name='Cluster', values = c("1" = "#e41a1c", "2"="#377eb8", "3"="#4daf4a"))   +  scale_y_log10(breaks = trans_breaks("log10", function(x) 10^x), labels = trans_format("log10", math_format(10^.x))) + guides(color='none')  +
  theme(axis.text.x = element_text(size=16), axis.text.y = element_text(size=16), axis.title.x = element_text(size=18), axis.title.y = element_text(size=18)) 

ggsave('./Analysis Figures/Figure1_subplots/Fig1-spRichness.svg', plotRichness, dpi=300, width=10, height=10)
ggsave('./Analysis Figures/Figure1_subplots/Fig1-spRicheness.png', plotRichness, dpi=300)

ggsave('./Analysis Figures/Figure1_subplots/Fig1-spDivShannon.svg', plotShannon, dpi=300, width=10, height=10)
ggsave('.//Analysis Figures/Figure1_subplots/Fig1-spDivShannon.png', plotShannon, dpi=300)

ggsave('./Analysis Figures/Figure1_subplots/Fig1-spDivFD.svg', plotFaithPD, dpi=300, width=10, height=10)
ggsave('.//Analysis Figures/Figure1_subplots/Fig1-spDivFD.png', plotFaithPD, dpi=300)


ggsave('./Analysis Figures/Figure1_subplots/Fig1-spDivSimpson.svg', plotSimpson, dpi=300, width=10, height=10)
ggsave('.//Analysis Figures/Figure1_subplots/Fig1-spDivSimpson.png', plotSimpson, dpi=300)

ggsave('./Analysis Figures/Figure1_subplots/Fig1-spDivSimpson_e.svg', plotSimpson_e, dpi=300, width=10, height=10)
ggsave('.//Analysis Figures/Figure1_subplots/Fig1-spDivSimpson_e.png', plotSimpson_e, dpi=300)

ggsave('./Analysis Figures/Figure1_subplots/Fig1-BactAbundance.svg', plotBacterialAmount, dpi=300, width=10, height=10)
ggsave('./Analysis Figures/Figure1_subplots/Fig1-BactAbundance.png', plotBacterialAmount, dpi=300)
```